src/charm.py

```diff
@@ -222,31 +253,30 @@ def _on_install(self, event: InstallEvent):
             event.defer()
             return
         self.unit.status = MaintenanceStatus("Initializing vault")
-        try:
-            (
-                private_key,
-                certificate,
-                ca_certificate,
-            ) = self._get_certificates_secret_in_peer_relation()
-        except PeerSecretError:
-            logger.info("Vault certificate secret not set in peer relation")
-            private_key, certificate, ca_certificate = generate_vault_certificates(
-                subject=self._certificate_subject,
-                sans_ip=[self._bind_address, self._ingress_address],
+        if not self._ca_certificate_pushed_to_workload():
+            ca_private_key, ca_certificate = generate_vault_ca_certificate()
+            self._set_ca_certificate_secret_in_peer_relation(
+                private_key=ca_private_key, certificate=ca_certificate
             )
-            self._set_certificates_secret_in_peer_relation(
-                private_key=private_key, certificate=certificate, ca_certificate=ca_certificate
+            self._push_ca_certificate_to_workload(certificate=ca_certificate)
+        if not self._unit_certificate_pushed_to_workload():
+            try:
+                ca_private_key, ca_certificate = self._get_ca_certificate_secret_in_peer_relation()
+            except PeerSecretError:
+                self.unit.status = WaitingStatus("Waiting for vault CA certificate secret")
+                event.defer()
+                return
+            sans_ip = [self._bind_address, self._ingress_address]
+            private_key, certificate = generate_vault_unit_certificate(
+                subject=self._ingress_address,
+                sans_ip=sans_ip,
+                ca_certificate=ca_certificate.encode(),
+                ca_private_key=ca_private_key.encode(),
             )
-        if not self._certificate_pushed_to_workload(
-            certificate=certificate,
-            private_key=private_key,
-            ca_certificate=ca_certificate,
-        ):
-            self._push_certificates_to_workload(
-                certificate=certificate,
-                private_key=private_key,
-                ca_certificate=ca_certificate,
+            self._push_unit_certificate_to_workload(
+                certificate=certificate, private_key=private_key
             )
+        self._delete_vault_data()
```

Comment:
why change the order of execution and delete the vault data after dealing with the certificates?

Reply:
I placed everything that could end up in deferring before so that we don't call it for no reason.